server/rest/src/main/java/org/infinispan/rest/resources/ProtobufResource.java

```diff
@@ -0,0 +1,233 @@
+   public Invocations getInvocations() {
+      return new Invocations.Builder()
+            // Key related operations
+            .invocation().methods(GET).path("/v2/schema").handleWith(this::getSchemasNames)
+            .invocation().methods(POST).path("/v2/schema/{schemaName}").handleWith(r -> addOrEditSchema(r,true))
+            .invocation().methods(PUT).path("/v2/schema/{schemaName}").handleWith(r -> addOrEditSchema(r,false))
+            .invocation().methods(GET).path("/v2/schema/{schemaName}").handleWith(this::getSchema)
+            .invocation().method(DELETE).path("/v2/schema/{schemaName}").handleWith(this::deleteSchema)
+            .create();
+   }
+
+   private CompletionStage<RestResponse> getSchemasNames(RestRequest request) {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+
+      AdvancedCache<Object, Object> cache = invocationHelper.getRestCacheManager()
+            .getCache(ProtobufMetadataManager.PROTOBUF_METADATA_CACHE_NAME, request);
+
+
+      return CompletableFuture.supplyAsync(() ->
+          Flowable.fromIterable(cache.entrySet())
+               .filter(cacheEntry -> !((String)cacheEntry.getKey()).endsWith(ProtobufMetadataManager.ERRORS_KEY_SUFFIX))
+               .map(cacheEntry -> {
+                  String error = (String) cache
+                        .get(cacheEntry.getKey() + ProtobufMetadataManagerConstants.ERRORS_KEY_SUFFIX);
+                  ProtoSchema protoSchema = new ProtoSchema();
+                  protoSchema.name = cacheEntry.getKey().toString();
+                  if (error != null) {
+                     protoSchema.error = createErrorContent(protoSchema.name, error);
+                  }
+                  return protoSchema;
+               })
+                .collectInto(new HashSet<ProtoSchema>(), Set::add)
+                .map(protoSchemas -> {
+                  try {
+                     List<ProtoSchema> sortedCacheInfos = protoSchemas.stream()
+                           .sorted(Comparator.comparing(s -> s.name))
+                           .collect(Collectors.toList());
+                     byte[] bytes = objectMapper.writeValueAsBytes(sortedCacheInfos);
+                     responseBuilder.contentType(APPLICATION_JSON).entity(bytes);
+                  } catch (JsonProcessingException e) {
+                     responseBuilder.status(HttpResponseStatus.INTERNAL_SERVER_ERROR);
+                  }
+                  return (RestResponse) responseBuilder.build();
+               })
+                .toCompletionStage()
+      , invocationHelper.getExecutor())
+            .thenCompose(Function.identity());
+   }
+
+   private CompletionStage<RestResponse> addOrEditSchema(RestRequest request, boolean create) {
+
+      String schemaName = checkMandatorySchemaName(request);
+
+      ContentSource contents = request.contents();
+      if (contents == null) throw new NoDataFoundException("Schema data not sent in the request");
```

Comment:
I am not sure how error handling is done on the rest side, but when returning a CompletionStage, I would normally expect this exception to be in the Stage.

Reply:
the exception is catchd in the BaseHttpRequestHandler
I'm following other examples
we should maybe talk about the overral design on this ?

Reply:
Sorry I don't see how BaseHttpRequestHandler can be possibly handling this. The handleError is only invoked from 2 places checking CompletionStage results. If a method can return a CompletionStage, it should _not_ be throwing an exception that is from user input. It should only be from bugs in our code.

Reply:
It is caught at RestDispatcherImpl#dispatch btw.

Reply:
can we deal this in another PR and talk about it with @gustavonalle ? because I'm not adding anything new and IMO this is another PR more global subject

Reply:
So I guess this code is okay, however it smells to me as we talked before at the F2F when the return type of a method is CompletionStage that any exceptions an invoker should be able to receive should be encapsulated in the Stage.

Reply:
BaseHttpRequestHandler#handleError should handle uncaught exceptions from the resources, and also log it
https://github.com/infinispan/infinispan/blob/1f25924b5d0a0f60e09d18004e50ea8fe54b42b4/server/rest/src/main/java/org/infinispan/rest/BaseHttpRequestHandler.java#L33

Could you please open a separate JIRA on how to align the error handling with what was discussed?

Reply:
is this JIRA not enough ? https://issues.redhat.com/browse/ISPN-11853